### ISSUE TITLE
patternProperties is broken when used by reference

### ISSUE THREAD

**Issue by luminoso** [koxudaxi/datamodel-code-generator#940]
**Describe the bug**
When `patternProperties` is inside a reference object, the generated classes are wrongly generated.

In this example, I created a json schema where:

- `KeyWithExplicitPatternProperties` has explicit `patternProperties`
- `KeyWithPatternPropertiesByReference` is the same as `KeyWithExplicitPatternProperties`, but it uses a `$ref` to define the same object as `KeyWithExplicitPatternProperties`
- `SomeOtherBoringReference` is just a boring `$ref` with a trivial object that is correctly parsed.

This way we can see that `KeyWithPatternPropertiesByReference` fails to correctly create the class. It looks related to the `$ref` invocation since everything else is the same.

To exemplify, I created a patternProperty that makes a json valid when the key is one letter and one char, like so:

```json

{
	"KeyWithExplicitPatternProperties": {
		"a1": "foo"
	},
	"KeyWithPatternPropertiesByReference": {
		"a2": "bar"
	},
	"SomeOtherBoringReference": {
		"timeout": 42
	}
}
```

**To Reproduce**

Example schema:
```json

{
  "$defs": {
    "stt": {
      "additionalProperties": false,
      "description": "STT properties allows to configure how the user voice is converted to text",
      "properties": {
        "timeout": {
          "title": "STT Timeout",
          "type": "number"
        }
      },
      "title": "Speech to text Settings",
      "type": "object"
    },
    "textResponse": {
      "additionalProperties": false,
      "type": "object",
      "patternProperties": {
        "^[a-z]{1}[0-9]{1}$": {
          "phoneNumber": {
            "additionalProperties": false,
            "type": "number"
          }
        }
      }
    }
  },
  "$schema": "http://json-schema.org/draft-07/schema",
  "additionalProperties": false,
  "description": "The root schema comprises the entire JSON document",
  "properties": {
    "KeyWithExplicitPatternProperties": {
      "additionalProperties": false,
      "type": "object",
      "patternProperties": {
        "^[a-z]{1}[0-9]{1}$": {
          "phoneNumber": {
            "additionalProperties": false,
            "type": "number"
          }
        }
      }
    },
    "KeyWithPatternPropertiesByReference": {
      "$ref": "#/$defs/textResponse"
    },
    "SomeOtherBoringReference": {
      "$ref": "#/$defs/stt"
    }
  },
  "title": "SomeSchema Schema",
  "type": "object"
}
```

Used commandline:
```
$ datamodel-codegen --use-double-quotes
```

**Expected behavior**

I would expect that `KeyWithPatternPropertiesByReference` is created the same way as `KeyWithExplicitPatternProperties`

**Version:**
 - OS: Fedora release 37 (Thirty Seven)
 - Python version: 3.10.8
 - datamodel-code-generator version: 0.14.0

**Additional context**

***Generated classes***

```python
# generated by datamodel-codegen:
#   filename:  v1.json
#   timestamp: 2022-12-13T16:49:55+00:00

from __future__ import annotations

from typing import Any, Dict, Optional

from pydantic import BaseModel, Extra, Field, constr


class Stt(BaseModel):
    class Config:
        extra = Extra.forbid

    timeout: Optional[float] = Field(None, title="STT Timeout")


class TextResponse(BaseModel):
    pass

    class Config:
        extra = Extra.forbid


class ConversationV1(BaseModel):
    class Config:
        extra = Extra.forbid

    KeyWithExplicitPatternProperties: Optional[Dict[constr(regex=r"^[a-z]{1}[0-9]{1}$"), Any]] = None
    KeyWithPatternPropertiesByReference: Optional[TextResponse] = None
    SomeOtherBoringReference: Optional[Stt] = None
```

To test, we can see that:

```python

valid_object1 = {"KeyWithExplicitPatternProperties": {"a1": "foo"}, "SomeOtherBoringReference": {"timeout": 42}}

conv1 = ConversationV1.parse_obj(valid_object1)

valid_object2 = {
    "KeyWithExplicitPatternProperties": {"a1": "foo"},
    "KeyWithPatternPropertiesByReference": {"a2": "bar"},
}

conv2 = ConversationV1.parse_obj(valid_object2)
```

`conv1` is correctly parsed, but `conv2` throws an Exception: 

```Traceback (most recent call last):
  File "/home/luminoso/scotty/bitbucket/stagger/src/gendatamodels/v1_gen.py", line 44, in <module>
    conv2 = ConversationV1.parse_obj(valid_object2)
  File "pydantic/main.py", line 526, in pydantic.main.BaseModel.parse_obj
  File "pydantic/main.py", line 342, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for ConversationV1
KeyWithPatternPropertiesByReference -> a2
  extra fields not permitted (type=value_error.extra)
```

Which is expected for the generated classes, but not for the schema.

I also validated the schema in https://www.jsonschemavalidator.net/

**Comment by koxudaxi** [comment]
@luminoso 
I have released a new version 0.17.0.
The problem has been fixed in the version.
Thank you very much!!

**Comment by luminoso** [comment]
Thank you! 😀
